File: client/tcp_client/control_client.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <cstring>
#include <vector>
#include <algorithm>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>

#define DEFAULT_PORT 2121
#define DEFAULT_IP "127.0.0.1"
#define BUFFER_SIZE 4096

class FTPClient {
private:
    int tcp_fd;
    bool connected;

    // Receive server response from TCP control channel
    std::string receive_response() {
        char buffer[BUFFER_SIZE];
        memset(buffer, 0, BUFFER_SIZE);

        ssize_t bytes_read = recv(tcp_fd, buffer, BUFFER_SIZE - 1, 0);

        if (bytes_read <= 0) {
            connected = false;
            return "";
        }

        return std::string(buffer);
    }

    // Send raw command string over TCP control connection
    bool send_command(const std::string& cmd) {
        std::string full_cmd = cmd + "\r\n";
        ssize_t bytes_sent = send(tcp_fd, full_cmd.c_str(), full_cmd.length(), 0);
        return bytes_sent > 0;
    }

    // Helper to extract response status code (e.g., 150, 220, 226)
    int get_response_code(const std::string& response) {
        if (response.length() < 3) return -1;
        try {
            return std::stoi(response.substr(0, 3));
        } catch (...) {
            return -1;
        }
    }

    // Placeholder/Handler for UDP Data transfer commands (LIST, RETR, STOR)
    void handle_data_transfer(const std::string& cmd_line) {
        // 1. Send the command over TCP control connection
        if (!send_command(cmd_line)) {
            std::cout << "[Error] Failed to send command to server.\n";
            return;
        }

        // 2. Read initial response (Expected: 150 File status okay / Opening data connection)
        std::string response = receive_response();
        std::cout << response;

        int code = get_response_code(response);
        if (code == 150) {
            // TODO: Execute UDP RDT protocol logic here
            // e.g., udp_client_receive_file() or udp_client_send_file()
            
            // 3. Read final completion response from server (Expected: 226 Transfer Complete)
            std::string final_res = receive_response();
            std::cout << final_res;
        }
    }

public:
    FTPClient() : tcp_fd(-1), connected(false) {}

    ~FTPClient() {
        disconnect();
    }

    // Connect to FTP server control port
    bool connect_to_server(const std::string& ip, int port) {
        tcp_fd = socket(AF_INET, SOCK_STREAM, 0);
        if (tcp_fd < 0) {
            std::cerr << "[Error] Failed to create socket!\n";
            return false;
        }

        sockaddr_in server_addr{};
        server_addr.sin_family = AF_INET;
        server_addr.sin_port = htons(port);

        if (inet_pton(AF_INET, ip.c_str(), &server_addr.sin_addr) <= 0) {
            std::cerr << "[Error] Invalid IP address format!\n";
            close(tcp_fd);
            return false;
        }

        if (connect(tcp_fd, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
            std::cerr << "[Error] Connection to server failed.\n";
            close(tcp_fd);
            return false;
        }

        connected = true;

        // Print welcome message (220)
        std::cout << receive_response();
        return true;
    }

    // Disconnect safely from FTP server
    void disconnect() {
        if (connected) {
            send_command("QUIT");
            std::cout << receive_response();
            close(tcp_fd);
            connected = false;
        }
    }

    // Main Interactive CLI Loop
    void run_cli() {
        std::string user_input;

        std::cout << "====================================================\n";
        std::cout << "   HYBRID FTP CLIENT CLI - READY FOR COMMANDS       \n";
        std::cout << "====================================================\n";
        std::cout << " [Auth/Session]: USER, PASS, QUIT\n";
        std::cout << " [Data Mode]   : PASV, PORT, TYPE\n";
        std::cout << " [Directory]   : PWD, CWD, CDUP, MKD, RMD\n";
        std::cout << " [File/Data]   : LIST, NLST, RETR, STOR, APPE, STOU, DELE, RNFR, RNTO, ABOR\n";
        std::cout << " [Info/Meta]   : STAT, SIZE, MDTM, HASH\n";
        std::cout << "====================================================\n\n";
        while (connected) {
            std::cout << "ftp> ";
            if (!std::getline(std::cin, user_input) || user_input.empty()) {
                continue;
            }

            // Extract command name in uppercase for comparison
            std::stringstream ss(user_input);
            std::string cmd;
            ss >> cmd;
            std::transform(cmd.begin(), cmd.end(), cmd.begin(), ::toupper);

            if (cmd == "EXIT" || cmd == "QUIT") {
                disconnect();
                break;
            }

            // Data transfer commands require two-step TCP response handling + UDP logic
            if (cmd == "LIST" || cmd == "NLST" || cmd == "RETR" || cmd == "STOR" || cmd == "APPE") {
                handle_data_transfer(user_input);
            } 
            else {
                // Standard control-only commands
                if (send_command(user_input)) {
                    std::string response = receive_response();
                    if (response.empty()) {
                        std::cout << "[Client] Server closed the connection.\n";
                        break;
                    }
                    std::cout << response;
                } else {
                    std::cout << "[Error] Failed to send command!\n";
                }
            }
        }
    }
};

int main(int argc, char* argv[]) {
    std::string ip = DEFAULT_IP;
    int port = DEFAULT_PORT;

    if (argc >= 2) ip = argv[1];
    if (argc >= 3) port = std::atoi(argv[2]);

    if (port <= 0 || port > 65535) {
        std::cerr << "[Fatal] Invalid port number: " << port << std::endl;
        return 1;
    }

    FTPClient client;
    std::cout << "[Client] Connecting to " << ip << ":" << port << "...\n";

    if (client.connect_to_server(ip, port)) {
        client.run_cli();
    }

    return 0;
}

File: client/tcp_client/control_client.py
import os
import sys
import socket

CURRENT_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_ROOT = os.path.abspath(os.path.join(CURRENT_DIR, "../../"))
if PROJECT_ROOT not in sys.path:
    sys.path.insert(0, PROJECT_ROOT)

from common.ftp_shared import (
    udp_client_receive_buffer,
    udp_client_receive_file,
    udp_client_send_file
)

DEFAULT_PORT = 2121
DEFAULT_IP = "127.0.0.1"
BUFFER_SIZE = 4096


class FTPClient:
    def __init__(self):
        self.tcp_sock: socket.socket | None = None
        self.connected: bool = False

    def _receive_response(self) -> str:
        if not self.tcp_sock:
            return ""
        try:
            data = self.tcp_sock.recv(BUFFER_SIZE)
            if not data:
                self.connected = False
                return ""
            return data.decode('utf-8', errors='ignore')
        except Exception:
            self.connected = False
            return ""

    def _send_command(self, cmd: str) -> bool:
        if not self.tcp_sock:
            return False
        full_cmd = cmd + "\r\n"
        try:
            bytes_sent = self.tcp_sock.send(full_cmd.encode('utf-8'))
            return bytes_sent > 0
        except Exception:
            return False

    @staticmethod
    def _get_response_code(response: str) -> int:
        if len(response) < 3:
            return -1
        try:
            return int(response[:3])
        except ValueError:
            return -1

    def handle_data_transfer(self, cmd_line: str) -> None:
        if not self._send_command(cmd_line):
            print("[Error] Failed to send command to server.")
            return

        response = self._receive_response()
        print(response, end="")

        code = self._get_response_code(response)
        if code == 150:
            parts = cmd_line.split(' ', 1)
            cmd = parts[0].upper()
            arg = parts[1].strip() if len(parts) > 1 else ""

            if cmd in ("LIST", "NLST"):
                data = udp_client_receive_buffer()
                if data:
                    print(data.decode('utf-8', errors='ignore'))
            elif cmd == "RETR":
                filename = arg if arg else "downloaded_file"
                udp_client_receive_file(filename)
            elif cmd in ("STOR", "APPE", "STOU"):
                udp_client_send_file(arg)

            final_res = self._receive_response()
            print(final_res, end="")

    def connect_to_server(self, ip: str, port: int) -> bool:
        try:
            self.tcp_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.tcp_sock.connect((ip, port))
            self.connected = True

            welcome_msg = self._receive_response()
            print(welcome_msg, end="")
            return True
        except Exception as e:
            print(f"[Error] Connection to server failed: {e}")
            if self.tcp_sock:
                self.tcp_sock.close()
            return False

    def disconnect(self) -> None:
        if self.connected:
            self._send_command("QUIT")
            print(self._receive_response(), end="")
            if self.tcp_sock:
                self.tcp_sock.close()
            self.connected = False

    def run_cli(self) -> None:
        print("====================================================")
        print("   HYBRID FTP CLIENT CLI - READY FOR COMMANDS       ")
        print("====================================================")
        print(" [Auth/Session]: USER, PASS, QUIT")
        print(" [Data Mode]   : PASV, PORT, TYPE")
        print(" [Directory]   : PWD, CWD, CDUP, MKD, RMD")
        print(" [File/Data]   : LIST, NLST, RETR, STOR, APPE, STOU, DELE, RNFR, RNTO, ABOR")
        print(" [Info/Meta]   : STAT, SIZE, MDTM, HASH")
        print("====================================================\n")

        try:
            while self.connected:
                user_input = input("ftp> ").strip()
                if not user_input:
                    continue

                parts = user_input.split()
                cmd = parts[0].upper()

                if cmd in ("EXIT", "QUIT"):
                    self.disconnect()
                    break

                if cmd in ("LIST", "NLST", "RETR", "STOR", "APPE", "STOU"):
                    self.handle_data_transfer(user_input)
                else:
                    if self._send_command(user_input):
                        response = self._receive_response()
                        if not response:
                            print("[Client] Server closed the connection.")
                            break
                        print(response, end="")
                    else:
                        print("[Error] Failed to send command!")
        except KeyboardInterrupt:
            print("\n[Client] Interrupted by user.")
            self.disconnect()


def main():
    ip = DEFAULT_IP
    port = DEFAULT_PORT

    if len(sys.argv) >= 2:
        ip = sys.argv[1]
    if len(sys.argv) >= 3:
        try:
            port = int(sys.argv[2])
        except ValueError:
            print(f"[Fatal] Invalid port number: {sys.argv[2]}")
            sys.exit(1)

    if port <= 0 or port > 65535:
        print(f"[Fatal] Invalid port number: {port}")
        sys.exit(1)

    client = FTPClient()
    print(f"[Client] Connecting to {ip}:{port}...")

    if client.connect_to_server(ip, port):
        client.run_cli()


if __name__ == "__main__":
    main()